File: CS/WebExample.Module.Web/MultiEditASPxGridListEditor.cs
using System;
using System.ComponentModel;
using System.Web.UI;

using DevExpress.ExpressApp.Web.Editors.ASPx;
using DevExpress.Web.ASPxGridView;
using DevExpress.ExpressApp.Editors;
using DevExpress.Web.ASPxClasses;
using DevExpress.ExpressApp.DC;
using DevExpress.Web.ASPxCallback;
using DevExpress.ExpressApp.Model;
using System.Web.UI.WebControls;
using DevExpress.ExpressApp.Web.Editors;

namespace WebExample.Module.Web {
    [ListEditor(typeof(object), false)]
    public class MultiEditASPxGridListEditor : ASPxGridListEditor {
        public MultiEditASPxGridListEditor(IModelListView model)
            : base(model) { }
        ASPxCallback callback;
        protected override object CreateControlsCore() {
            Panel panel = new Panel();
            callback = new ASPxCallback();
            callback.ID = Model.Id + "aspxCallback1";
            callback.ClientInstanceName = Model.Id + "_callback1";
            callback.Callback += new CallbackEventHandler(callback_Callback);
            panel.Controls.Add(callback);
            ASPxGridView grid = (ASPxGridView)base.CreateControlsCore();
            panel.Controls.Add(grid);
            return panel;
        }
        protected override ITemplate CreateDataItemTemplate(IModelColumn columnInfo) {
            if (IsColumnSupported(columnInfo)) {
                WebPropertyEditor propertyEditor = FindPropertyEditor(columnInfo, ViewEditMode.Edit);
                propertyEditor.ImmediatePostData = false;
                return new MultiRowEditTemplate(propertyEditor, callback.ClientInstanceName);
            } else {
                return base.CreateDataItemTemplate(columnInfo);
            }
        }
        void callback_Callback(object source, DevExpress.Web.ASPxCallback.CallbackEventArgs e) {
            String[] p = e.Parameter.Split('|');
            Object key = TypeDescriptor.GetConverter(ObjectTypeInfo.KeyMember.MemberType).ConvertFromString(p[0]);
            IMemberInfo member = ObjectTypeInfo.FindMember(p[1]);
            Object value = TypeDescriptor.GetConverter(member.MemberType).ConvertFromString(p[2]); ;
            object obj = ObjectSpace.GetObjectByKey(ObjectTypeInfo.Type, key);
            member.SetValue(obj, value);
            ObjectSpace.CommitChanges();
        }
        Type[] supportedPropertyEditorTypes = new Type[]{
            typeof(ASPxStringPropertyEditor),
            typeof(ASPxIntPropertyEditor),
            typeof(ASPxBooleanPropertyEditor),
            typeof(ASPxEnumPropertyEditor)
        };
        protected virtual bool IsColumnSupported(IModelColumn model) {
            if (model.GroupIndex >= 0) {
                return false;
            }
            foreach (Type type in supportedPropertyEditorTypes) {
                if (type.IsAssignableFrom(model.PropertyEditorType)) {
                    return true;
                }
            }
            return false;
        }
        // Sorting and grouping are not supported
        protected override ColumnWrapper AddColumnCore(IModelColumn columnInfo) {
            ASPxGridViewColumnWrapper columnWrapper = (ASPxGridViewColumnWrapper)base.AddColumnCore(columnInfo);
            if (IsColumnSupported(columnInfo)) {
                columnWrapper.Column.Settings.AllowSort = DevExpress.Utils.DefaultBoolean.False;
                columnWrapper.Column.Settings.AllowGroup = DevExpress.Utils.DefaultBoolean.False;
            }
            return columnWrapper;
        }
    }
    public class MultiRowEditTemplate : IBehaviourTemplate {
        WebPropertyEditor propertyEditor;
        string callbackName;
        const String CallbackArgumentFormat = "function (s, e) {{ {0}.PerformCallback(\"{1}|{2}|\" + {3}); }}"; // ASPxCallback, key, fieldName, value

        public MultiRowEditTemplate(WebPropertyEditor propertyEditor, string callbackName) {
            this.propertyEditor = propertyEditor;
            this.callbackName = callbackName;
        }
        #region IBehaviourTemplate Members

        public void InstantiateIn(Control container) {
            GridViewDataItemTemplateContainer gridContainer = container as GridViewDataItemTemplateContainer;
            if (gridContainer == null) {
                throw new NotSupportedException(container.GetType().FullName);
            }
            propertyEditor.CreateControl();
            Object obj = gridContainer.Grid.GetRow(gridContainer.VisibleIndex);
            propertyEditor.CurrentObject = obj;
            propertyEditor.ReadValue();
            if (propertyEditor.Editor is ASPxWebControl) {
                ((ASPxWebControl)propertyEditor.Editor).SetClientSideEventHandler("ValueChanged", String.Format(CallbackArgumentFormat,
                    callbackName, gridContainer.KeyValue, gridContainer.Column.FieldName, "s.GetValue()"));
            }
            container.Controls.Add((Control)propertyEditor.Control);
        }

        public bool CancelClickEventPropagation {
            get {
                return true;
            }
        }

        #endregion
    }
}

File: VB/WebExample.Module.Web/MultiEditASPxGridListEditor.vb
Imports System
Imports System.ComponentModel
Imports System.Web.UI

Imports DevExpress.ExpressApp.Web.Editors.ASPx
Imports DevExpress.Web.ASPxGridView
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.Web.ASPxClasses
Imports DevExpress.ExpressApp.DC
Imports DevExpress.Web.ASPxCallback
Imports DevExpress.ExpressApp.Model
Imports System.Web.UI.WebControls
Imports DevExpress.ExpressApp.Web.Editors

Namespace WebExample.Module.Web
    <ListEditor(GetType(Object), False)> _
    Public Class MultiEditASPxGridListEditor
        Inherits ASPxGridListEditor
        Public Sub New(ByVal model As IModelListView)
            MyBase.New(model)
        End Sub
        Private callback As ASPxCallback
        Protected Overrides Function CreateControlsCore() As Object
            Dim panel As New Panel()
            callback = New ASPxCallback()
            callback.ID = Model.Id & "aspxCallback1"
            callback.ClientInstanceName = Model.Id & "_callback1"
            AddHandler callback.Callback, AddressOf callback_Callback
            panel.Controls.Add(callback)
            Dim grid As ASPxGridView = CType(MyBase.CreateControlsCore(), ASPxGridView)
            panel.Controls.Add(grid)
            Return panel
        End Function
        Protected Overrides Function CreateDataItemTemplate(ByVal columnInfo As IModelColumn) As ITemplate
            If IsColumnSupported(columnInfo) Then
                Dim propertyEditor As WebPropertyEditor = FindPropertyEditor(columnInfo, ViewEditMode.Edit)
                propertyEditor.ImmediatePostData = False
                Return New MultiRowEditTemplate(propertyEditor, callback.ClientInstanceName)
            Else
                Return MyBase.CreateDataItemTemplate(columnInfo)
            End If
        End Function
        Private Sub callback_Callback(ByVal source As Object, ByVal e As DevExpress.Web.ASPxCallback.CallbackEventArgs)
            Dim p() As String = e.Parameter.Split("|"c)
            Dim key As Object = TypeDescriptor.GetConverter(ObjectTypeInfo.KeyMember.MemberType).ConvertFromString(p(0))
            Dim member As IMemberInfo = ObjectTypeInfo.FindMember(p(1))
            Dim value As Object = TypeDescriptor.GetConverter(member.MemberType).ConvertFromString(p(2))

            Dim obj As Object = ObjectSpace.GetObjectByKey(ObjectTypeInfo.Type, key)
            member.SetValue(obj, value)
            ObjectSpace.CommitChanges()
        End Sub
        Private supportedPropertyEditorTypes() As Type = { GetType(ASPxStringPropertyEditor), GetType(ASPxIntPropertyEditor), GetType(ASPxBooleanPropertyEditor), GetType(ASPxEnumPropertyEditor) }
        Protected Overridable Function IsColumnSupported(ByVal model As IModelColumn) As Boolean
            If model.GroupIndex >= 0 Then
                Return False
            End If
            For Each type As Type In supportedPropertyEditorTypes
                If type.IsAssignableFrom(model.PropertyEditorType) Then
                    Return True
                End If
            Next type
            Return False
        End Function
        ' Sorting and grouping are not supported
        Protected Overrides Function AddColumnCore(ByVal columnInfo As IModelColumn) As ColumnWrapper
            Dim columnWrapper As ASPxGridViewColumnWrapper = CType(MyBase.AddColumnCore(columnInfo), ASPxGridViewColumnWrapper)
            If IsColumnSupported(columnInfo) Then
                columnWrapper.Column.Settings.AllowSort = DevExpress.Utils.DefaultBoolean.False
                columnWrapper.Column.Settings.AllowGroup = DevExpress.Utils.DefaultBoolean.False
            End If
            Return columnWrapper
        End Function
    End Class
    Public Class MultiRowEditTemplate
        Implements IBehaviourTemplate
        Private propertyEditor As WebPropertyEditor
        Private callbackName As String
        Private Const CallbackArgumentFormat As String = "function (s, e) {{ {0}.PerformCallback(""{1}|{2}|"" + {3}); }}" ' ASPxCallback, key, fieldName, value

        Public Sub New(ByVal propertyEditor As WebPropertyEditor, ByVal callbackName As String)
            Me.propertyEditor = propertyEditor
            Me.callbackName = callbackName
        End Sub
        #Region "IBehaviourTemplate Members"

        Public Sub InstantiateIn(ByVal container As Control) Implements ITemplate.InstantiateIn
            Dim gridContainer As GridViewDataItemTemplateContainer = TryCast(container, GridViewDataItemTemplateContainer)
            If gridContainer Is Nothing Then
                Throw New NotSupportedException(container.GetType().FullName)
            End If
            propertyEditor.CreateControl()
            Dim obj As Object = gridContainer.Grid.GetRow(gridContainer.VisibleIndex)
            propertyEditor.CurrentObject = obj
            propertyEditor.ReadValue()
            If TypeOf propertyEditor.Editor Is ASPxWebControl Then
                CType(propertyEditor.Editor, ASPxWebControl).SetClientSideEventHandler("ValueChanged", String.Format(CallbackArgumentFormat, callbackName, gridContainer.KeyValue, gridContainer.Column.FieldName, "s.GetValue()"))
            End If
            container.Controls.Add(CType(propertyEditor.Control, Control))
        End Sub

        Public ReadOnly Property CancelClickEventPropagation() As Boolean Implements IBehaviourTemplate.CancelClickEventPropagation
            Get
                Return True
            End Get
        End Property

        #End Region
    End Class
End Namespace